Moving CSI Driver to GA.

huffmanca opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Graduates the CSI Driver to GA. This only touches the API, their tests, and the associated staging files.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/enhancements/issues/770

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
The storage.k8s.io/CSIDriver has moved to GA, and is now available for use.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/768
```

gnufied commented:
/ok-to-test
/sig storage
/priority important-soon

jennybuckley commented:
/remove-sig api-machinery

jsafrane commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+// CSI drivers do not need to create the CSIDriver object directly. Instead they may use the
+// cluster-driver-registrar sidecar container. When deployed with a CSI driver it automatically
```
We do not support cluster-driver-registrar any longer: https://kubernetes-csi.github.io/docs/cluster-driver-registrar.html

jsafrane reviewed (commented):
* I think you should also flip the feature gate to GA and remove any code that flips the feature gate to false (basically only in unit tests).
* A/D controller and kubelet should use storage/v1 CSIDriver instead of v1beta1. Basically nothing in kubelet and kube-controller-manager should import `k8s.io/api/storage/v1beta1` (at least not for CSIDriver).

https://github.com/kubernetes/kubernetes/pull/83474 is a good example to follow.

jsafrane commented:
> *I think you should also flip the feature gate to GA and remove any code that flips the feature gate to false (basically only in unit tests).

\+ Jordan says "drop all associated checks of ...FeatureGate.Enabled(features.CSIDriver) and assume they evaluate to true now" https://github.com/kubernetes/kubernetes/pull/83568/files#r344719331

bertinatto commented on pkg/volume/csi/csi_plugin.go:
```diff
@@ -807,6 +763,7 @@ func (p *csiPlugin) supportsVolumeLifecycleMode(driver string, volumeMode storag
 		return fmt.Errorf("volume mode %q not supported by driver %s (only supports %q)", volumeMode, driver, csiDriver.Spec.VolumeLifecycleModes)
 	}
 }
+*/
```
Uncomment?

bertinatto commented on pkg/features/kube_features.go:
```diff
@@ -561,7 +561,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	ServiceNodeExclusion:           {Default: false, PreRelease: featuregate.Alpha},
 	NodeDisruptionExclusion:        {Default: false, PreRelease: featuregate.Alpha},
-	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.Beta},
+	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.GA},
```
LockToDefault: true

huffmanca commented:
/test pull-kubernetes-node-e2e-containerd

huffmanca commented:
/test pull-kubernetes-node-e2e-containerd

huffmanca commented:
/test pull-kubernetes-node-e2e-containerd

huffmanca commented:
/test pull-kubernetes-e2e-gce-100-performance

huffmanca commented:
/test pull-kubernetes-node-e2e-containerd

huffmanca commented:
/test pull-kubernetes-node-e2e-containerd

huffmanca commented:
/retest

huffmanca commented:
/retest

huffmanca commented:
/retest

huffmanca committed:
Moved CSIDriver to GA

huffmanca committed:
Updated CSIDriver references

smarterclayton commented:
/assign @smarterclayton 

for api review

huffmanca commented:
/retest

gnufied committed:
Fix expected version for csidriver

Add a comment about GA feature

gnufied commented:
/retest

jsafrane commented on pkg/registry/storage/rest/storage_storage.go:
```diff
@@ -144,6 +144,15 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
 	}
 
+	// register csidrivers if CSIDriverRegistry feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSIDriverRegistry) {
```
The feature is enabled and can't be disabled, no need to check for it.

huffmanca replied:
I've removed these checks wherever they were found. I noticed that there are still some checks for other features (such as CSINodeInfo), should we create an issue to remove these checks as well?

jsafrane commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
+	// https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html
+	// A driver can support one or more of these modes and
+	// more modes may be added in the future.
+	// +optional
```
Please add a note that this field is beta.

jsafrane reviewed (commented):
All `features.CSIDriverRegistry` checks should be removed from the code, the feature should be always on
(see https://github.com/kubernetes/kubernetes/pull/84814/files#r385148181)

huffmanca commented:
/test pull-kubernetes-e2e-gce-alpha-features

msau42 commented on pkg/features/kube_features.go:
```diff
@@ -585,7 +585,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	ServiceNodeExclusion:           {Default: false, PreRelease: featuregate.Alpha},
 	NodeDisruptionExclusion:        {Default: false, PreRelease: featuregate.Alpha},
-	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.GA},
+	CSIDriverRegistry:              {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```
Add a comment to remove in 1.20

huffmanca replied:
Comment has been included.

msau42 commented on test/integration/etcd/data.go:
```diff
@@ -503,13 +503,17 @@ func GetEtcdStorageDataForNamespace(namespace string) map[schema.GroupVersionRes
+	// k8s.io/kubernetes/pkg/apis/storage/v1
+	etcdStorageData[gvr("storage.k8s.io", "v1", "csidrivers")] = StorageData{
+		Stub:             `{"metadata": {"name": "csid2"}, "spec": {"attachRequired": true, "podInfoOnMount": true}}`,
+		ExpectedEtcdPath: "/registry/csidrivers/csid2",
```
We need to keep the default GVK to be v1beta for 1 release to handle HA master version skew.

See [here](https://github.com/kubernetes/kubernetes/pull/83474/commits/603a2aa8a99869dc51ea1be2ce6e6ca10c23e55f#diff-a6932c2bd6398cd55df3be4564939a5eR559) and [here](https://github.com/kubernetes/kubernetes/pull/83474/commits/603a2aa8a99869dc51ea1be2ce6e6ca10c23e55f#diff-c74f8b85666a09e8b8a2a052f4a91abdR57)

msau42 replied:
Oops, I see it's added in the 3rd commit. Would still be good to add a TODO and open an issue to track removing it next release.

huffmanca replied:
I've updated this to include a TODO, and created issue https://github.com/kubernetes/kubernetes/issues/88821 to track it.

gnufied commented:
/label api-review
/assign @msau42 @jsafrane

huffmanca commented:
/test pull-kubernetes-e2e-kind

huffmanca commented:
/retest

huffmanca commented:
/retest

huffmanca commented:
/retest

huffmanca commented:
/retest

jsafrane commented:
/lgtm
/approve
pkg/volume and other storage parts (i.e. not API)

huffmanca commented:
/retest

jsafrane commented:
snapshots tests are flaky and unrelated to this PR (https://github.com/kubernetes/kubernetes/pull/88801)

huffmanca commented:
/retest

saad-ali commented:
/milestone v1.18

msau42 commented:
Minor nit on release note: Clarify it's the storage.k8s.io/CSIDriver object. It currently reads as if it's talking about some vendor csi driver.

huffmanca commented:
/test pull-kubernetes-e2e-kind

huffmanca commented:
/test pull-kubernetes-e2e-gce

huffmanca commented:
> Minor nit on release note: Clarify it's the storage.k8s.io/CSIDriver object. It currently reads as if it's talking about some vendor csi driver.

The release note has been updated.

smarterclayton commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
+	// +optional
+	PodInfoOnMount *bool `json:"podInfoOnMount,omitempty" protobuf:"bytes,2,opt,name=podInfoOnMount"`
+
+	// VolumeLifecycleModes defines what kind of volumes this CSI volume driver supports.
```
This should be `volumeLifecycleModes` - always reference the external name in external API doc.

huffmanca replied:
This has been updated with the latest push.

smarterclayton commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -221,6 +221,133 @@ type VolumeError struct {
+	// https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html
+	// A driver can support one or more of these modes and
+	// more modes may be added in the future.
+	// This field is beta.
```
This is not actually true after this PR?  Or is it?

smarterclayton replied:
Isn't there another PR in flight that changes the description of this field?  If so, you need to incorporate those changes before I'll approve it (I *think* it got approved yesterday).

huffmanca replied:
The `volumeLifecycleModes` field is still beta, while the rest of the types should be GA

gnufied replied:
Did a PR change description of `volumeLifecycle` mode recently? I might have missed it. @smarterclayton which PR you are referring to? cc @msau42

huffmanca replied:
I looked and wasn't able to find a PR that changed this description, either open or recently merged.

smarterclayton reviewed (changes requested):
Some final things that need to be cleared up

huffmanca commented:
/test pull-kubernetes-e2e-kind

huffmanca commented:
@smarterclayton ,

After discussing this with @gnufied  further, I've removed `listType` from the Items[] field and added a single API violation for this type.

smarterclayton commented:
/approve

API change is approved.

huffmanca commented:
/test pull-kubernetes-e2e-kind

msau42 commented on staging/src/k8s.io/api/storage/v1/types.go:
```diff
@@ -221,6 +221,131 @@ type VolumeError struct {
+	// For more information about implementing this mode, see
+	// https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html
+	// A driver can support one or more of these modes and
+	// more modes may be added in the future.
```
I don't see the note that this field is beta?

gnufied replied:
It is there below your comment.

msau42 replied:
i swear it wasn't there..... ok :-)

msau42 commented:
/lgtm

huffmanca commented:
/test pull-kubernetes-e2e-gce

irvifa commented:
@huffmanca I think you need to rebase this PR first.. There's a conflict on the `csi_mounter`

huffmanca committed:
Updated files after rebase

huffmanca commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

jsafrane commented:
/lgtm
